src/coprocessor/dag/expr/mod.rs

```diff
@@ -258,45 +303,83 @@ impl TryFrom<Expr> for Expression {
                 .map(Datum::Dec)
                 .map(|e| Expression::new_const(e, tp))
                 .map_err(Error::from),
-            // TODO(andelf): fn sig verification
             ExprType::ScalarFunc => {
-                let sig = expr.get_sig();
-                let mut expr = expr;
+                try!(FnCall::check_args(
+                    expr.get_sig(),
+                    expr.get_children().len()
+                ));
                 expr.take_children()
                     .into_iter()
-                    .map(Expression::try_from)
+                    .map(|child| Expression::build(child, row_len))
                     .collect::<Result<Vec<_>>>()
                     .map(|children| {
                         Expression::ScalarFn(FnCall {
-                            sig: sig,
+                            sig: expr.get_sig(),
                             children: children,
                             tp: tp,
                         })
                     })
             }
-            ExprType::ColumnRef => expr.get_val()
-                .decode_i64()
-                .map(|i| {
-                    Expression::ColumnRef(Column {
-                        offset: i as usize,
-                        tp: tp,
-                    })
-                })
-                .map_err(Error::from),
+            ExprType::ColumnRef => {
+                let offset = try!(expr.get_val().decode_i64().map_err(Error::from)) as usize;
+                try!(Column::check_offset(offset, row_len));
+                let column = Column {
+                    offset: offset,
+                    tp: tp,
+                };
+                Ok(Expression::ColumnRef(column))
+            }
             unhandled => unreachable!("can't handle {:?} expr in DAG mode", unhandled),
         }
     }
 }
 
-#[test]
-fn test_smoke() {
-    use std::convert::TryInto;
-    use util::codec::number::NumberEncoder;
-
-    let mut pb = Expr::new();
-    pb.set_tp(ExprType::ColumnRef);
-    pb.mut_val().encode_i64(1).unwrap();
+#[cfg(test)]
+mod test {
+    use coprocessor::codec::Datum;
+    use coprocessor::select::xeval::evaluator::test::{col_expr, datum_expr};
+    use tipb::expression::{Expr, ExprType, FieldType, ScalarFuncSig};
+    use super::Expression;
+
+    pub fn fncall_expr(sig: ScalarFuncSig, ft: FieldType, children: &[Expr]) -> Expr {
+        let mut expr = Expr::new();
+        expr.set_tp(ExprType::ScalarFunc);
+        expr.set_sig(sig);
+        expr.set_field_type(ft);
+        for child in children {
+            expr.mut_children().push(child.clone());
+        }
+        expr
+    }
 
-    let e: Result<Expression> = pb.try_into();
-    let _ = e.unwrap();
+    #[test]
+    fn test_expression_build() {
```

Comment:
now there are two failure cases: one for columnOffset, and one for bad function arguments.